kong/conf_loader.lua

```diff
@@ -560,6 +565,47 @@ local function check_and_infer(conf)
     errors[#errors + 1] = "router_update_frequency must be greater than 0"
   end
 
+  if conf.role == "control_plane" then
+    if #conf.admin_listen < 1 or pl_stringx.strip(conf.admin_listen[1]) == "off" then
+      errors[#errors + 1] = "admin_listen must be specified when role = \"control_plane\""
+    end
+
+    if #conf.cluster_listen < 1 or pl_stringx.strip(conf.cluster_listen[1]) == "off" then
+      errors[#errors + 1] = "cluster_listen must be specified when role = \"control_plane\""
+    end
+
+    if conf.database == "off" then
+      errors[#errors + 1] = "in-memory storage can not be used when role = \"control_plane\""
+    end
+
+  elseif conf.role == "data_plane" then
+    if #conf.proxy_listen < 1 or pl_stringx.strip(conf.proxy_listen[1]) == "off" then
+      errors[#errors + 1] = "proxy_listen must be specified when role = \"data_plane\""
+    end
+
+    if conf.database ~= "off" then
```

Comment:
Should we just force this `off` when role is `data_plane`, and document that with `data_plane` role the `database` configuration is not used?